src/policy/fees.cpp

```diff
@@ -76,116 +78,109 @@ void TxConfirmStats::UpdateMovingAverages()
+    // convert from satoshis to COIN
+    minTxFee = minTxFee / COIN;
+
     // Counters for a bucket (or range of buckets)
     double nConf = 0; // Number of tx's confirmed within the confTarget
     double totalNum = 0; // Total number of tx's that were ever confirmed
-    int extraNum = 0; // Number of tx's still in mempool for confTarget or longer
-
-    int maxbucketindex = buckets.size() - 1;
-
-    // requireGreater means we are looking for the lowest fee/priority such that all higher
-    // values pass, so we start at maxbucketindex (highest fee) and look at successively
-    // smaller buckets until we reach failure.  Otherwise, we are looking for the highest
-    // fee/priority such that all lower values fail, and we go in the opposite direction.
-    unsigned int startbucket = requireGreater ? maxbucketindex : 0;
-    int step = requireGreater ? -1 : 1;
-
-    // We'll combine buckets until we have enough samples.
-    // The near and far variables will define the range we've combined
-    // The best variables are the last range we saw which still had a high
-    // enough confirmation rate to count as success.
-    // The cur variables are the current range we're counting.
-    unsigned int curNearBucket = startbucket;
-    unsigned int bestNearBucket = startbucket;
-    unsigned int curFarBucket = 0;
-    unsigned int bestFarBucket = startbucket;
-
-    bool foundAnswer = false;
-    unsigned int bins = unconfTxs.size();
-
-    // Start counting from highest(default) or lowest fee/pri transactions
-    for (int bucket = startbucket; bucket >= 0 && bucket <= maxbucketindex; bucket += step)
+    int32_t extraNum = 0; // Number of tx's still in mempool for confTarget or longer
+
+    int32_t maxbucketindex = buckets.size() - 1;
+
+    // We want a sub-vector of buckets to be our range from which we select a bucket for our fee
+    // so instead of copying each bucket into a new vector we can simulate a sub-vector that is
+    // between bucketFront and bucketBack (naming from std::vector.front and std::vector.back)
+    int32_t selectedBucket = -1;
+
+    uint32_t bins = unconfTxs.size();
+
+    // Start counting from highest fee transactions
+    for (int32_t bucket = maxbucketindex; bucket >= 0 && bucket <= maxbucketindex; bucket = bucket - 1)
     {
-        curFarBucket = bucket;
+        // add the moving average number of confirmed tx's for the conf target in bucket
         nConf += confAvg[confTarget - 1][bucket];
+        // add the moving average number of transactions in bucket to the total number of transactions
         totalNum += txCtAvg[bucket];
-        for (unsigned int confct = confTarget; confct < GetMaxConfirms(); confct++)
+        for (uint32_t confct = confTarget; confct < GetMaxConfirms(); confct++)
+        {
+            // add number of unconfirmed transactions for a conf target in the given bucket (less than MAX_CONFIRMS)
             extraNum += unconfTxs[(nBlockHeight - confct) % bins][bucket];
```

Comment:
I investigated for 5 minutes and I'm not completely convinced that (nBlockHeight - confct) % bins > 0 in all weird conditions.  We need to either eliminate this possibility or check for it.

Reply:
the mod of a number is never negative. N mod M will always produce a natural number between 0 and M -1